x-pack/plugins/endpoint/server/routes/endpoints.ts

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { IRouter } from 'kibana/server';
+import { SearchResponse } from 'elasticsearch';
+import { schema } from '@kbn/config-schema';
+import { EndpointAppContext, EndpointData } from '../types';
+import { AllEndpointsQueryBuilder } from '../services/endpoint/endpoint_query_builders';
+
+interface HitSource {
+  _source: EndpointData;
+}
+
+export interface EndpointResultList {
+  endpoints: EndpointData[];
+  total: number;
+  requestPageSize: number;
+  requestIndex: number;
+}
+
+export function registerEndpointRoutes(router: IRouter, endpointAppContext: EndpointAppContext) {
+  router.get(
+    {
+      path: '/api/endpoint/endpoints',
+      validate: {
+        query: schema.object({
+          pageSize: schema.number({ defaultValue: 10, min: 1 }),
+          pageIndex: schema.number({ defaultValue: 0, min: 0 }),
+        }),
+      },
+      options: { authRequired: true },
+    },
+    async (context, req, res) => {
+      try {
+        const queryParams = await new AllEndpointsQueryBuilder(
+          req,
+          endpointAppContext
+        ).toQueryParams();
+        const response = (await context.core.elasticsearch.dataClient.callAsCurrentUser(
+          'search',
+          queryParams
+        )) as SearchResponse<EndpointData>;
+        return res.ok({ body: await mapToEndpointResultList(queryParams, response) });
+      } catch (err) {
+        return res.internalError({ body: err });
+      }
+    }
+  );
+}
+
+function mapToEndpointResultList(
+  queryParams: Record<string, unknown>,
+  searchResponse: SearchResponse<EndpointData>
+): EndpointResultList {
+  if (searchResponse.hits.hits.length > 0) {
+    return {
+      requestPageSize: queryParams.size as number,
```

Comment:
Would you explain why you are casting values in this function? Thanks

Reply:
It fails on type check hence it was cast.

Reply:
I think since `queryParams` is typed as `Record<string, unknown>`, the type system doesn't understand that `queryParams.size` is a number. Maybe you could try changing that type to `queryParams: { size: number, from: number }`. This function relies on those values being numbers, so the type could reflect that.

Reply:
turn out it was not needed, but in case you are wondering why it is typed as such https://github.com/elastic/kibana/blob/master/src/core/server/elasticsearch/scoped_cluster_client.ts#L67. I changed it to `Record<string, any>` which is the type of the for the client call